### PR TITLE
Avoid locks and allocations in the common path.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs [hunk 1]
```diff
@@ -80,9 +80,24 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
             var metadata = GetMetadataNoThrow(reference);
             if (metadata == null)
             {
-                return null;
+                return SpecializedTasks.Default<SymbolTreeInfo>();
+            }
+
+            // Try to acquire the data outside the lock.  That way
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs]: That way what?

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide to take a look.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs [hunk 2]
```diff
@@ -80,9 +80,24 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
             var metadata = GetMetadataNoThrow(reference);
             if (metadata == null)
             {
-                return null;
+                return SpecializedTasks.Default<SymbolTreeInfo>();
+            }
+
+            // Try to acquire the data outside the lock.  That way 
+            SymbolTreeInfo info;
+            if (s_metadataIdToInfo.TryGetValue(metadata.Id, out info))
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs]: Is this the only place s_metadataIdToInfo is used?  If so, consider storing the Task already in the dictionary, as this still allocates on every `Task.FromResult` call.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs [hunk 3]
```diff
@@ -80,9 +80,27 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
...
+            }
+
+            // Try to acquire the data outside the lock.  That way we can avoid any sort of 
+            // allocations around acquiring the task for it.  Note: once ValueTask is available
+            // (and enabled in the language), we'd likely want to use it here. (Presuming 
+            // the lock is not being held most of the time).
+            SymbolTreeInfo info;
+            if (s_metadataIdToInfo.TryGetValue(metadata.Id, out info))
+            {
+                return Task.FromResult(info);
```
Comment (heejaechang) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs]: this still allocates though.

Comment (Pilchie) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs]: Why not update `s_metadataIdToInfo` to store a `Task<SymbolTreeInfo>`, so you can give back the existing task every time.

**Comment by asecchia** [comment]
Type System.Threading.Tasks.Task`1[Roslyn.Utilities.SemaphoreSlimExtensions+SemaphoreDisposer] 	 0.0	441,616
500MB -> 441KB

**Pull request merged**